File: Code/Cpp/Factory/Factory.cpp
#include <string>
#include <iostream>

using namespace std;

class Train {
    private: 
    int id;
    string color;

    public:
    Train(int id, string color){
        this->id = id;
        this->color = color;
    }
     void describe() {
        printf("I am a train with id = %d and color = %s\n",id, color.c_str());
    }
};

class ClassicTrain : public Train {
    public: ClassicTrain(int id) : Train(id, "red") {}
};

class ModernTrain : public Train {
    public: ModernTrain(int id) : Train(id, "white") {}
};

class TrainCreator {
    public: virtual Train createTrain();
};

class ModernTrainCreator {
    public: Train createTrain(){
        return *(new ModernTrain(1));
    }
};

class ClassicTrainCreator {
    public: Train createTrain(){
        return *(new ClassicTrain(2));
    }
};

int main(){
    ModernTrainCreator* modernTrainCreator = new ModernTrainCreator();
    ClassicTrainCreator* classicTrainCreator = new ClassicTrainCreator();

    modernTrainCreator->createTrain().describe();
    classicTrainCreator->createTrain().describe();
}

File: Code/Python/Factory/Factory.py
from abc import ABC, abstractmethod

class Train(ABC):
    @abstractmethod
    def __init__(self, id, color):
        self.__trainID = id
        self.__color = color

    def describe(self):
        print("I am a train with id = %s and color = %s" % (self.__trainID, self.__color))

class ClassicTrain(Train):
    def __init__(self, id):
        super().__init__(id, "red")

class ModernTrain(Train):
    def __init__(self, id):
        super().__init__(id, "white")

class TrainCreator(ABC):
    @abstractmethod
    def createTrain(self):
        pass

class ModernTrainCreator(TrainCreator):
    def createTrain(self):
        return ModernTrain(1)

class ClassicTrainCreator(TrainCreator):
    def createTrain(self):
        return ClassicTrain(2)        

modernTrain = ModernTrainCreator().createTrain().describe()
classicTrain = ClassicTrainCreator().createTrain().describe()